### DISCUSSION TITLE
A faster way to perform proximity analysis

### DISCUSSION THREAD

**Discussion by cristiangaldini** [NetTopologySuite/NetTopologySuite#693] [Q&A]
Hi, basically i am working with a collection of points and a collection of lines.
My goal is to identify the nearest lines from each node (around a maximum distance ).

I tested various approaches based on buffering the node and test intersections, like

var prepBuffer = NetTopologySuite.Geometries.Prepared.PreparedGeometryFactory.Prepare(geom);
foreach (var geomItem in lines)
{
   
    if (prepBuffer .Intersects(geomItem))
    {
        // do 
    }
}
or 
using
CreateGeometryCollection.Intersect

or based on calculating distances, using also IsWithinDistance.

In anycase using a set of data consisting on 80000 lines and 300000 points, the performaces are very poor (tens of minutes on last 
generation icore7)

Is there any different efficient way to implement this algorithm?
thanks
regards
Cristian

**Comment by tommygebhardt** [comment]
You can use a spatial index like the STRtree to store the nodes. For the Envelope you can expand the point envelope by your maximum distance. Then when you query the index, you can test each point in the returned results for the minimum distance

**Reply by cristiangaldini** [reply]
thanks, i will try

**Reply by tommygebhardt** [reply]
Thinking of this a little more, I would not expand the envelopes of the point. I would instead expand the search envelope. Fewer operations. I would also expand the search envelope from the line envelope, rather than buffering the line and then getting the envelope. It's less precise but much faster - and all you're doing with that is shrinking your search space for the distance measurements

**Reply by cristiangaldini** [reply]
Hi, Sure i used the search envelop from point. The STRtree is a  STRtree of LineString , then
foreach (NetTopologySuite.Geometries.Geometry pc in gcPoints.Geometries)
    var pbuf = pc.Buffer(10);
                var res= gpsSTRtree.Query(new Envelope(pbuf.Coordinates));
but uhm..,.
do you thinking to use STRtree of points and a query envelope buffering the linestring?
i suppose is not correct, i could find a point near the linestring, but that point could be nearest to another linestring

btw, found the line, i can identify the point on line nearest to examinated point ? ( the traight line projection)

**Reply by tommygebhardt** [reply]
If you are looking for just the nearest line to a point, this code should work:
```c#
IEnumerable<(Point p, LineString l)> NearestLinesToPoints(IEnumerable<Point> points, IEnumerable<LineString> lines)
{
    STRtree<Geometry> rtree = new STRtree<Geometry>();
    foreach( LineString line in lines)
    {
        rtree.Insert(line.EnvelopeInternal, line);
    }
    rtree.Build();

    var geoDist = new GeometryItemDistance();
    foreach( Point p in points )
    {
        var nearestLine = (LineString) rtree.NearestNeighbour(p.EnvelopeInternal, p, geoDist);
        yield return (p, nearestLine);
    }
}
``` 

If you're looking for all of the lines within a distance of a point, then instead of the NearestNeighbor search you would do something like
```c#
foreach( Point p in points )
  {
      var searchEnv = p.EnvelopeInternal;
      searchEnv.ExpandBy(threshold);
      var nearbyLines = rtree.Query(searchEnv);
      var linesWithinThresh = nearbyLines.Cast<LineString>().Where(l => p.Distance(l) < threshold);
      yield return (p, linesWithinThresh);
  }
```

In response to the point on line nearest the estimated point, you can use the DistanceToPoint class:
```c#
PointPairDistance dist = new PointPairDistance();
dist.Initialize();
DistanceToPoint.ComputeDistance(line, point.Coordinate, dist);
var closestCoord = dist.Coordinates[0];
```

**Reply by cristiangaldini** [reply]
wow, these codes are very fast, around 10 seconds to finish the job...

thanks a lot

  var linesWithinThresh =  nearbyLines.Cast<LineString>().Where(l => p.Distance(l) < threshold);
return a list, so we can identify the nearest from the list

      var line_nearest_WithinThresh =  nearbyLines.Cast<LineString>().Where(l => p.Distance(l) < threshold).OrderBy(l => p.Distance(l)).FirstOrDefault();

btw, if we query by an Envelope expanded by threshold, why we have to test again distance < threshold?

**Reply by tommygebhardt** [reply]
The rtree query isn't exactly performing the distance check. It's (roughly) looking for envelopes that intersect the search envelope. So it's quite possible you can get a feature that is outside of the threshold.

For instance, you have a `POINT (0,0)` and a `LINESTRING(5 -1, -1 5)`. The features are at minimum 2√2 apart . If your search distance was 1, your envelope would be `Env(-1 -1 , 1 1)`  which would intersect the lines envelope which is `Env(-1 -1 , 5 5)`.

If your max distance is fuzzy, you may not care. But if it's a hard limit, you need a distance check after the query.

**Reply by cristiangaldini** [reply]
i see ,thanks
